File: app/lib/features/favorites/data/favorites_repository.dart
import 'package:sbc_contacts/core/network/api_client.dart';
import 'package:sbc_contacts/core/network/paginated.dart';
import 'package:sbc_contacts/features/directory/domain/member.dart';

/// A favorited member (backend `FavoriteItem`) mapped onto [Member] for reuse.
Member _favoriteToMember(Map<String, dynamic> json) => Member(
      id: (json['memberSbcId'] ?? '').toString(),
      sbcId: (json['memberSbcId'] ?? '').toString(),
      name: json['name'] as String?,
      firstName: json['firstName'] as String?,
      profession: json['profession'] as String?,
      city: json['city'] as String?,
      country: json['country'] as String?,
      avatarUrl: json['avatarUrl'] as String?,
      phoneNumber: json['phoneNumber'] as String?,
      isFavorite: true,
    );

class FavoritesRepository {
  FavoritesRepository(this._api);
  final ApiClient _api;

  Future<Paginated<Member>> list({int page = 1, int limit = 50}) async {
    final data = await _api.get('/favorites', query: {'page': page, 'limit': limit})
        as Map<String, dynamic>;
    return Paginated.fromJson(data, _favoriteToMember);
  }

  Future<void> add(String memberSbcId) =>
      _api.post('/favorites', body: {'memberSbcId': memberSbcId});

  Future<void> remove(String memberSbcId) => _api.delete('/favorites/$memberSbcId');
}

File: ios/SBCContacts/Features/Favorites/FavoritesRepository.swift
import Foundation

/// A favorited member (backend `FavoriteItem`), mapped onto `Member` for reuse.
struct FavoriteItem: Decodable, Sendable {
    let member: Member

    private enum CodingKeys: String, CodingKey {
        case memberSbcId, name, firstName, profession, city, country, avatarUrl, phoneNumber
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        let sbcId = c.string(.memberSbcId) ?? ""
        member = Member(
            id: sbcId,
            sbcId: sbcId,
            name: c.string(.name),
            firstName: c.string(.firstName),
            profession: c.string(.profession),
            city: c.string(.city),
            country: c.string(.country),
            avatarUrl: c.string(.avatarUrl),
            phoneNumber: c.string(.phoneNumber),
            isFavorite: true
        )
    }
}

struct FavoritesRepository: Sendable {
    let api: APIClient

    private struct AddBody: Encodable, Sendable {
        let memberSbcId: String
    }

    func list(page: Int = 1, limit: Int = 50) async throws -> Paginated<Member> {
        let result: Paginated<FavoriteItem> = try await api.get(
            "/favorites",
            query: [.init(name: "page", value: String(page)), .init(name: "limit", value: String(limit))]
        )
        return result.map(\.member)
    }

    func add(memberSbcId: String) async throws {
        try await api.post("/favorites", body: AddBody(memberSbcId: memberSbcId))
    }

    func remove(memberSbcId: String) async throws {
        try await api.delete("/favorites/\(memberSbcId)")
    }
}
